practice/rain.py

```diff
@@ -0,0 +1,193 @@
+"""Rain database processor.
+
+Accesses City of Portland rain total database and returns various information
+about rain totals.
+"""
+import urllib.request
+import re
+import os.path
+import statistics
+
+
+def get_daily_totals(filename):
+    """Return a dict of days and daily amounts, as rows from the file."""
+    totals = {}
+    with open(filename) as f:
+        for i in range(11):
+            _ = f.readline()
+        for line in f:
+            day_stats = line.split()[0:2]
+            if day_stats[1] != '-':  # skip entries if there is missing data
+                totals.update({day_stats[0]: int(day_stats[1])})
+    return totals
+
+
+def find_wettest_stats(dates_totals):
+    """Return wettest day (or year) and the amount on the wettest day (or year).
+
+    >>> find_wettest_stats({'30-MAR-2016': 0, '29-MAR-2016': 5})
+    ('29-MAR-2016', 5)
+    """
+    totals_dates = {v: k for k, v in dates_totals.items()}
+    largest = max(totals_dates)
+    return totals_dates[largest], largest
+
+
+def calc_yearly_total_stats(totals):
+    """Return a dictionary of years and sum amounts.
+
+    >>> sorted(calc_yearly_total_stats({'30-MAR-2016': 1, '29-MAR-2016': 5,
+    ...                         '30-MAR-2015': 0, '29-MAR-2015': 5}).items())
+    [('2015', 5), ('2016', 6)]
+    """
+    yearly_totals = {}
+    for date_data, amt in totals.items():
+        year_in_date_data = date_data.split('-')[2]
+        if year_in_date_data not in yearly_totals:
+            yearly_totals.update({year_in_date_data: 0})
+        yearly_totals[year_in_date_data] += amt
+    return yearly_totals
```

Comment:
For `calc_yearly_total_stats()`, try using `groupby()` and splitting out the key calculation into a function.

I want to encourage people to use high-level transformations as opposed to verbosely writing out all instructions.